File: c4_linkedlist/L2-m.cpp
//
// Created by LiHeng on 2020/4/19.
//

#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {    //加法十进制向后进位
        ListNode* head = new ListNode(-1);   //存放结果的链表
        ListNode* h = head;//移动指针
        int sum = 0;//每个位的加和结果
        bool carry = false;//进位标志
        while(l1 != NULL || l2 != NULL) //采用||处理后，只要有一个没走到结尾就不会退出循环，
             // 已经走到结尾的在循环体内部再做处理，这样就不需要处理数字位数不一样而在短的后面补0
        {
            sum = 0;
            if(l1 != NULL) {
                sum += l1->val;
                l1 = l1->next;
            }
            if(l2 != NULL) {
                sum += l2->val;
                l2 = l2->next;
            }
            if(carry) //是否加1判定;依据上一步是否进位的判定标志carry
                sum++;
            h->next = new ListNode(sum%10);//创建新节点
            h = h->next;
            carry = sum>=10 ? true:false;
        }
        if(carry)
            h->next = new ListNode(1); //处理最高位进1
        return head->next;
    }
};

File: python-algorithm/ds_4_linkedlist/linkedlist_1_vertical_calculation/L2-m.py
#--coding:utf-8--

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0)
        tmp = dummy
        carry = 0
        while l1 or l2:
            t_sum = 0
            if l1:
                t_sum += l1.val
                l1 = l1.next
            if l2:
                t_sum += l2.val
                l2 = l2.next
            t_sum += carry
            cur_nd = ListNode(t_sum%10)
            carry = t_sum//10
            tmp.next = cur_nd
            tmp = tmp.next
        if carry:
            cur_nd = ListNode(carry)
            tmp.next = cur_nd
            tmp = tmp.next
        return dummy.next

